### PR TITLE
Temporal robustness

### PR DESCRIPTION
**Pull request by srikeerthi207**
temporal robustness - added continuous and iterative models

### CODE REVIEW

File: docs/temporal_robustness/continuous_model.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
+import os
+# export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/home/username/anaconda3/lib/
```
Comment (jasonjewik) [inline docs/temporal_robustness/continuous_model.py:2]: Remove this commented-out line.

File: docs/temporal_robustness/continuous_model.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+os.environ['LD_LIBRARY_PATH']
```
Comment (jasonjewik) [inline docs/temporal_robustness/continuous_model.py:3]: What is the purpose of this line?

File: docs/temporal_robustness/continuous_model.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+from climate_learn.utils.datetime import Year, Days, Hours
+from climate_learn.data import DataModule
+from climate_learn.models import set_climatology
+from climate_learn.models import load_model
+from climate_learn.data import IterDataModule
+from climate_learn.training import Trainer, WandbLogger
+from climate_learn.models import load_model
+import torch
+
+torch.cuda.empty_cache()
```
Comment (jasonjewik) [inline docs/temporal_robustness/continuous_model.py:15]: What is this line for?

File: docs/temporal_robustness/continuous_model.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+pred_range = Hours(6)
+
+forecast_data_module = IterDataModule(
+    task="forecasting",
+    inp_root_dir="/data0/datasets/weatherbench/data/weatherbench/era5/5.625deg_npz",
```
Comment (jasonjewik) [inline docs/temporal_robustness/continuous_model.py:21]: Please do not expose the directory structure of our cluster.

File: docs/temporal_robustness/continuous_model.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    logger = WandbLogger(project = project_name, name = "era"),
+    early_stopping=False,
+    task = "forecasting"
+)
+
+forecast_trainer.fit(forecast_model_module, forecast_data_module)
+
+forecast_trainer.test(forecast_model_module, forecast_data_module)
+
+# forecast_trainer.test(forecast_model_module, forecast_data_module, "path")
```
Comment (jasonjewik) [inline docs/temporal_robustness/continuous_model.py:74]: Please remove this commented line.

File: docs/temporal_robustness/direct_model.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
+import os
+# export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/home/username/anaconda3/lib/
```
Comment (jasonjewik) [inline docs/temporal_robustness/direct_model.py:2]: See my comments on `continuous_model.py`. Also, why are these files in the docs folder?

File: src/climate_learn/models/modules/utils/metrics.py [hunk 7]
```diff
@@ -2,11 +2,10 @@
 import torch
 from scipy import stats
 
-
 ### Training loss
 
 
-def mse(pred, y, vars, lat=None, log_postfix=""):
+def mse(pred, y, vars, lat=None, log_postfix="", splice_out_variables = -1):
```
Comment (jasonjewik) [inline src/climate_learn/models/modules/utils/metrics.py:8]: What is `splice_out_variables` doing?

File: src/climate_learn/training/train.py [hunk 8]
```diff
@@ -71,5 +71,9 @@ def __init__(
     def fit(self, model_module, data_module):
         self.trainer.fit(model_module, data_module)
 
-    def test(self, model_module, data_module):
-        self.trainer.test(model_module, data_module)
+    def test(self, model_module, data_module, checkpoint_path=""):
```
Comment (jasonjewik) [inline src/climate_learn/training/train.py:74]: IMO if `checkpoint_path` is an optional argument, it should default to `None`, not the empty string.

File: src/climate_learn/training/train.py [hunk 9]
```diff
@@ -71,5 +71,9 @@ def __init__(
...
+        if checkpoint_path:
+            print("checkpoint_path", checkpoint_path)
```
Comment (jasonjewik) [inline src/climate_learn/training/train.py:76]: Printing seems unnecessary to me.

**Review by jasonjewik — CHANGES_REQUESTED** [review]
PR not approved.
1. It seems there is a merge conflict that needs to be resolved.
2. This PR depends upon #79, which has not yet been approved.
3. Please include examples of the new features you have added. I suggest creating a tutorial notebook in the `notebooks` folder.

File: src/climate_learn/models/modules/utils/metrics.py (hunk 7 above)
Comment (srikeerthi207) [reply src/climate_learn/models/modules/utils/metrics.py:8]: for iterative models, in the intermediate steps we predict all the 36 output vars, where as in the final test step we predict only the required 3 output vars. This splice_out_variables is used in the evaluation step to compute the final metrics on the 3 output vars

File: docs/temporal_robustness/continuous_model.py (hunk 3 above)
Comment (srikeerthi207) [reply docs/temporal_robustness/continuous_model.py:15]: not required, I will remove it

**Comment by srikeerthi207** [comment]
> PR not approved.
> 
> 1. It seems there is a merge conflict that needs to be resolved.
> 2. This PR depends upon #79, which has not yet been approved.
> 3. Please include examples of the new features you have added. I suggest creating a tutorial notebook in the `notebooks` folder.

I created .py files for each of the three models (direct_model.py, continuous_model.py, iterative_model.py) instead of a notebook, so that it can be run on a screen

File: src/climate_learn/models/__init__.py [hunk 10]
```diff
@@ -5,7 +5,7 @@
 from climate_learn.utils.datetime import Hours
 
 
-def load_model(name, task, model_kwargs, optim_kwargs):
+def load_model(name, task, model_kwargs, optim_kwargs, iterative_model = False, iter_model_out_vars = -1):
```
[the remark this answers is not in the document]
Comment (srikeerthi207) [reply src/climate_learn/models/__init__.py:8]: for iterative models, 3 output variables should be predicted (and the metrics are also supposed to be computed on 3 output vars) in the test step of final iteration. But, the intermediate steps training will compute all 36 output vars and compute the validation step metrics on all 36 output vars. So by passing as model_kwargs, we can't distinguish between the  evaluate call of validation and test steps.

**Comment by jasonjewik** [comment]
> I created .py files for each of the three models (direct_model.py, continuous_model.py, iterative_model.py) instead of a notebook, so that it can be run on a screen

Right, so this is useful for experiments, but for documentation/tutorial purposes, having these in some sort of Jupyter notebook would be nice. However, we can shelve this for now until we decide what to do with the new model arguments (see [this comment](https://github.com/aditya-grover/climate-learn/pull/84#discussion_r1146390477)).

**Comment by prakhar6sharma** [comment]
@jasonjewik are we going ahead with this PR? or are we leaving #88 to implement continuous and iterative model?

**Pull request closed without merging**